.github/workflows/test.yml

```
@@ -18,18 +18,18 @@ jobs:
     container:
       image: ubuntu:focal
     steps:
-    - uses: actions/checkout@v2
+    - uses: actions/checkout@v3
```

version `v4` is available

```suggestion
    - uses: actions/checkout@v4
```

same below

---

you need to change "node" permissions to remove "node" version 12